microsoft/testsuites/cloud_hypervisor/ch_tests_tool.py

```diff
@@ -0,0 +1,159 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+import json
+from dataclasses import dataclass
+from pathlib import PurePath
+from typing import Any, List, Optional, Type
+
+from assertpy.assertpy import assert_that
+
+from lisa import Environment, notifier
+from lisa.executable import Tool
+from lisa.messages import CommunityTestMessage, TestStatus, create_test_result_message
+from lisa.operating_system import CBLMariner
+from lisa.testsuite import TestResult
+from lisa.tools import Docker, Echo, Git, Whoami
+
+
+@dataclass
+class CloudHypervisorTestResult:
+    name: str = ""
+    status: TestStatus = TestStatus.QUEUED
+
+
+class CloudHypervisorTests(Tool):
+    TIME_OUT = 3600
+
+    repo = "https://github.com/cloud-hypervisor/cloud-hypervisor.git"
+
+    cmd_path: PurePath
+    repo_root: PurePath
+
+    @property
+    def command(self) -> str:
+        return str(self.cmd_path)
+
+    @property
+    def can_install(self) -> bool:
+        return True
+
+    @property
+    def dependencies(self) -> List[Type[Tool]]:
+        return [Git, Docker]
+
+    def run_tests(
+        self,
+        test_result: TestResult,
+        environment: Environment,
+        test_type: str,
+        skip: Optional[List[str]] = None,
+    ) -> None:
+        if skip is not None:
+            skip_args = " ".join(map(lambda t: f"--skip {t}", skip))
+        else:
+            skip_args = ""
+
+        result = self.run(
+            f"tests --{test_type} -- -- {skip_args} -Z unstable-options --format json",
+            timeout=self.TIME_OUT,
+            force_run=True,
+            cwd=self.repo_root,
+            no_info_log=False,  # print out result of each test
+            shell=True,
+        )
+
+        results = self._extract_test_results(result.stdout)
+        failures = list(filter(lambda r: r.status == TestStatus.FAILED, results))
+        if not failures:
+            result.assert_exit_code()
+
+        for r in results:
+            self._send_community_test_msg(
+                test_result.id_,
+                environment,
+                r.name,
+                r.status,
+            )
+
```

Comment:
does it mean only when all cases pass, then they can be inserted into db?

Reply:
No. Even when there are some failures and some successes, all the results will be inserted into the DB.

If you are referring to the `result.assert_exit_code()` above, that is just to ensure that the script didn't fail due to some reason other than test failures. The test script does a lot of pre-requisite steps before actually executing the tests and things can potentially fail here. If there are no test failures and still the exit code is non-zero that means something has gone wrong with script execution. In this case we will just fail the main testcase.